### ISSUE TITLE
PHPCS freezes when run with anything < -vv

### ISSUE THREAD

**Issue by simPod** [squizlabs/PHP_CodeSniffer#2659]
_Using Macos and recently upgraded to Catalina. Catalina asks for every tiny permission so it might be related somehow. I'm not certain though but timeline of this issue and os upgrade is similar._

- `vendor/bin/phpcs` – process is running, cpu high, never ends
- `vendor/bin/phpcs -v` – process is running, cpu high, never ends
- `vendor/bin/phpcs -vv` – process is running, cpu high, finishes in reasonable time
- `vendor/bin/phpcs -vvv` – process is running, cpu high, finishes in reasonable time

There's something that `-vv` flag does as it probably allows the process not to stuck in a loop.

Ideas what might be wrong appreciated, I'm going to dig into sniffer code.

**Comment by simPod** [comment]
With `-v` the output is

```
Registering sniffs in the  standard... DONE (145 sniffs registered)
Creating file list... DONE (311 files in queue)
Loading cache... DONE (311 files in cache)
```
cleaning the cache has no effect

**Comment by simPod** [comment]
It's the `$this->config->parallel = 1;` that makes it work.

**Comment by simPod** [comment]
When parallel `2` is used, 2 tmp files are created and filled with content, final tmp file is also getting filled. The correct content is already inside of `phpcss5mvqv` but all processes are still running for few hours now.

**Comment by gsherwood** [comment]
As you've found, this is obviously an issue with running checks in parallel. Using `-vv` disables this so it can print accurate debug output.

Although I don't know why this would not be working.

> When parallel `2` is used, 2 tmp files are created and filled with content, final tmp file is also getting filled. The correct content is already inside of `phpcss5mvqv` but all processes are still running for few hours now.

Are you saying that the child processes are writing the file with their summary output here:
https://github.com/squizlabs/PHP_CodeSniffer/blob/ae3aae46a11e3978f088552160ccca13a01b28e1/src/Runner.php#L513-L538

**Comment by simPod** [comment]
@gsherwood yes and it's also getting picket ny main process it seems. Have to investigate further.

**Comment by simPod** [comment]
It's not easy to debug but it seems like `$res = pcntl_waitpid($procData['pid'], $status, WNOHANG);` always returns 0 event though child process called `exit()` already. I'll try with php7.2 (7.3 now)

**Comment by simPod** [comment]
Ok, on 7.2.23 `pcntl_waitpid` returns actual PID. On 7.3.10 always 0. Using macos brew php distribution. I suppose it's out of scope of this lib then.

**Comment by jrfnl** [comment]
@simPod Noticed you opened a bug report upstream :+1: 

For reference if anyone else comes across this issue: https://bugs.php.net/bug.php?id=78717

**Comment by simPod** [comment]
@jrfnl I wanted to link it here but I'm still not sure what is the cause. Though it's macos issue but tried on another mac and it works there. However, I can easily reproduce it by this script https://gist.github.com/simPod/5a177947bdf864026bfc59d71d6e9d71

**Comment by jrfnl** [comment]
@simPod Unfortunately I'm on Windows, so can't try to reproduce this (PCNTL is not available on Windows). I hope someone can help figure it out. Have you been able to determine in which PHP version it was introduced ? I.e. 7.3.0 or a later patch version ?

**Comment by simPod** [comment]
@jrfnl I tried to install 7.3.6 which had the other mac device where it was working. No success, still looping infinitely. This will be fun to discover the cause.
